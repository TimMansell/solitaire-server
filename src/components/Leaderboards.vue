<template>
  <div data-test="leaderboards">
    <p v-if="name" data-test="leaderboard-name">
      Your player name is: {{ name }}
    </p>

    <div ref="scrollTo">
      <Filters>
        <Select
          v-model="filters.showBest"
          label="Best"
          :items="['moves', 'times']"
          @select="setBest"
          data-test="leaderboard-set-best"
        />

        <Select
          v-model="filters.limit"
          label="Top"
          :items="['25', '50', '100', '500']"
          @select="displayLimit"
          data-test="leaderboard-set-top"
        />
      </Filters>
    </div>

    <p data-test="leaderboards-heading">Top {{ limit }} Best {{ best }}</p>

    <ResponsiveTable
      :headings="['Rank', 'Date', 'Player', `${best}`]"
      :items="leaderboards"
      :placeholder-rows="limit"
      :to-highlight="{ key: 'player', value: name }"
    />
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';
import Filters from '@/components/Filters.vue';
import Select from '@/components/Select.vue';
import ResponsiveTable from '@/components/ResponsiveTable.vue';

export default {
  name: 'Leaderboards',
  components: {
    Filters,
    Select,
    ResponsiveTable,
  },
  data() {
    return {
      filters: {
        limit: parseInt(this.$route.params.limit, 10),
        showBest: this.$route.params.showBest,
      },
    };
  },
  watch: {
    filters: {
      async handler() {
        await this.displayGames();

        this.updateUrl();
        this.scrollTo();
      },
      deep: true,
    },
  },
  computed: {
    ...mapGetters(['leaderboards', 'name']),
    showBest() {
      return this.filters.showBest;
    },
    limit() {
      return this.filters.limit;
    },
    best() {
      const { showBest } = this.filters;

      return showBest.charAt(0).toUpperCase() + showBest.slice(1);
    },
  },
  mounted() {
    this.displayGames();
  },
  methods: {
    ...mapActions(['getLeaderboards']),
    displayLimit(limit) {
      this.filters.limit = parseInt(limit, 10);
    },
    setBest(showBest) {
      this.filters.showBest = showBest;
    },
    async displayGames() {
      const { filters } = this;

      await this.getLeaderboards(filters);
    },
    scrollTo() {
      this.$emit('scrollTo', this.$refs.scrollTo);
    },
    updateUrl() {
      const { showBest, limit } = this;

      this.$router.replace(`/leaderboards/${showBest}/${limit}`);
    },
  },
};
</script>
