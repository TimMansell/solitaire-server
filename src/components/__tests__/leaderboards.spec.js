import { shallowMount } from '@vue/test-utils';
import Leaderboards from '@/components/Leaderboards.vue';

const mocks = {
  $store: { dispatch: jest.fn() },
};

const mockUid = 'f5c6a829-f0da-4dfc-81a0-e6419f0163c7';

const mockLeaderboardsMoves = [
  {
    rank: 1,
    date: '2021-04-29T12:25:47.907Z',
    uid: '7dac9d78-353f-409b-8a7f-2192409c44a2',
    moves: 2,
  },
  {
    rank: 2,
    date: '2021-04-29T12:26:20.825Z',
    uid: '2cbf658a-3102-4e9d-b749-bac853efed0d',
    moves: 2,
  },
];

const mockComputed = {
  leaderboards: () => mockLeaderboardsMoves,
  luid: () => mockUid,
};

describe('Leaderboards.vue', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(Leaderboards, {
      mocks,
      computed: {
        ...mockComputed,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });
});
