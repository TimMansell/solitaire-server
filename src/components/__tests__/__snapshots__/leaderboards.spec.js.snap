// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboards.vue matches snapshot 1`] = `
<div data-test="leaderboards" class="leaderboards">
  <p data-test="leaderboard-name">Your player name is: f5c6a829-f0da-4dfc-81a0-e6419f0163c7</p>
  <div>
    <filters-stub>
      <select-stub label="Best" value="Moves" items="Moves,Times" data-test="leaderboard-set-best"></select-stub>
      <select-stub label="Top" value="25" items="25,50,100,500" data-test="leaderboard-set-top"></select-stub>
    </filters-stub>
  </div>
  <p data-test="leaderboards-heading">Top 25 Best Moves</p>
  <responsivetable-stub headings="Rank,Date,Player,Moves" items="[object Object],[object Object]" placeholderrows="25" tohighlight="[object Object]"></responsivetable-stub>
</div>
`;
